slurm_ops_manager/nvidia.py

```diff
@@ -0,0 +1,167 @@
+
+    @property
+    def installed(self) -> bool:
+        """Return wether nvidia driver is installed."""
+        return self._stored.gpu_installed
+
+    @property
+    def repository(self):
+        """Return the repository used for nvidia drivers."""
+        if self._stored.gpu_repo_configured:
+            return Path(self._stored.gpu_repo_path).read_text()
+        else:
+            return "Repository not set up"
+
+    @repository.setter
+    def repository(self, repo: str):
+        """Set a custom repository to install nvidia drivers.
+
+        If the new repository string is empty, re-sets to the default.
+        """
+        if repo:
+            logger.debug(f"#### Nvidia - setting custom repo: {repo}")
+            Path(self._stored.gpu_repo_path).write_text(repo)
+        else:
+            logger.debug("#### Nvidia - setting default repo")
+            # instructions from nvdia website:
+            # https://docs.nvidia.com/datacenter/tesla/tesla-installation-notes/index.html
+
+            if self._operating_system == "ubuntu":
+                distro = "ubuntu2004"
+                repo = f"deb http://developer.download.nvidia.com/compute/cuda/repos/{distro}/x86_64 /" # noqa
+                Path(self._stored.gpu_repo_path).write_text(repo)
+
+                # download pin file
+                url = f"https://developer.download.nvidia.com/compute/cuda/repos/{distro}/x86_64/cuda-{distro}.pin" # noqa
+                path = "/etc/apt/preferences.d/nvidia_gpu-pin-600"
+                cmd = f"curl -s -L -o {path} {url}"
+                subprocess.run(shlex.split(cmd))
+
+                # add key
+                cmd = f"apt-key adv --fetch-keys https://developer.download.nvidia.com/compute/cuda/repos/{distro}/x86_64/7fa2af80.pub" # noqa
+                subprocess.run(shlex.split(cmd))
+
+                # it is necessary to update apt's db after every repo change
+                subprocess.call(["apt-get", "update"])
+            elif self._operating_system == "centos":
+                distro = "rhel7"
+                url = f"http://developer.download.nvidia.com/compute/cuda/repos/{distro}/x86_64/cuda-{distro}.repo" # noqa
+                cmd = f"yum-config-manager --add-repo {url}"
+                subprocess.run(shlex.split(cmd))
+
+                subprocess.call(["yum", "clean", "expire-cache"])
+
+        self._stored.gpu_repo_configured = True
+
+    @property
+    def package(self) -> str:
+        """Return nVdia package name."""
+        return self._stored.gpu_package_name
+
+    @package.setter
+    def package(self, pkg: str):
+        """Change the package to install."""
+        if not pkg:
+            # if empty, sets the default
+            if self._operating_system == "ubuntu":
+                self._stored.gpu_package_name = "cuda-drivers"
+            elif self._operating_system == "centos":
+                self._stored.gpu_package_name = "nvidia-driver-latest-dkms cuda cuda-drivers"
```

Comment:
Not sure I am following you on this. Can you expand a bit?

Reply:
In other places where you are checking the operating system, you have an `else` clause that indicates that an unsupported operating system is being used and you log the error. I think it would make sense to do that here as well.

Reply:
This part of the code will never be run if the operating is not CentOS7 or Ubuntu 20.04. We had support for older Ubuntu versions, but we dropped them. In some places, this `else` clause was a guard against that. But it is not needed anymore.